src/System.Net.Security/src/System/Net/SecureProtocols/_SslStream.cs

```diff
@@ -0,0 +1,709 @@
+            if (_InternalBufferFromPinnableCache)
+            {
+                s_PinnableReadBufferCache.FreeBuffer(_InternalBuffer);
+                _InternalBufferFromPinnableCache = false;
+            }
+            _InternalBuffer = null;
+        }
+
+        ~_SslStream()
+        {
+            if (_InternalBufferFromPinnableCache)
+            {
+                if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                {
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Read Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_InternalBuffer));
+                }
+                FreeReadBuffer();
+            }
+            if (_PinnableOutputBuffer != null)
+            {
+                if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                {
+                    PinnableBufferCacheEventSource.Log.DebugMessage2("DTOR: In System.Net._SslStream.~SslStream Freeing Write Buffer", this.GetHashCode(), PinnableBufferCacheEventSource.AddressOfByteArray(_PinnableOutputBuffer));
+                }
+                s_PinnableWriteBufferCache.FreeBuffer(_PinnableOutputBuffer);
+            }
+        }
+
+        internal int Read(byte[] buffer, int offset, int count)
+        {
+            return ProcessRead(buffer, offset, count, null);
+        }
+
+        internal void Write(byte[] buffer, int offset, int count)
+        {
+            ProcessWrite(buffer, offset, count, null);
+        }
+
+        internal bool DataAvailable
+        {
+            get { return InternalBufferCount != 0; }
+        }
+
+        private byte[] InternalBuffer
+        {
+            get
+            {
+                return _InternalBuffer;
+            }
+        }
+
+        private int InternalOffset
+        {
+            get
+            {
+                return _InternalOffset;
+            }
+        }
+
+        private int InternalBufferCount
+        {
+            get
+            {
+                return _InternalBufferCount;
+            }
+        }
+
+        private void DecrementInternalBufferCount(int decrCount)
+        {
+            _InternalOffset += decrCount;
+            _InternalBufferCount -= decrCount;
+        }
+
+        //
+        // This will set the internal offset to "curOffset" and ensure internal buffer.
+        // If not enough, reallocate and copy up to "curOffset".
+        //
+        private void EnsureInternalBufferSize(int curOffset, int addSize)
+        {
+            if (_InternalBuffer == null || _InternalBuffer.Length < addSize + curOffset)
+            {
+                bool wasPinnable = _InternalBufferFromPinnableCache;
+                byte[] saved = _InternalBuffer;
+
+                int newSize = addSize + curOffset;
+                if (newSize <= PinnableReadBufferSize)
+                {
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
```

Comment:
Related to `PinnableBufferCache` @brianrob may be able to explain.
Ideally, UTs should exist to determine the answer and to pin behavior.